encoding/text/marshal.go

```diff
@@ -453,3 +544,65 @@ func (ew *errWriter) WriteByte(b byte) error {
 	}
 	return ew.err
 }
+
+// indentWriter is helper for writing indented text
+type indentWriter struct {
+	w errWriter
+
+	// how many spaces to indent each output level with, 0 disables indentation
+	indentPerLevel int
+
+	// current indent level
+	currentIndent int
+
+	// hasLineContent is true when we have written something on the current line.
+	hasLineContent bool
+}
+
+func (iw *indentWriter) beforeWrite() {
+	if iw.indentPerLevel > 0 && !iw.hasLineContent {
+		iw.hasLineContent = true
+		spaces := iw.indentPerLevel * iw.currentIndent
+		for i := 0; i < spaces; i++ {
+			iw.w.WriteByte(' ')
+		}
+	}
+}
+
+func (iw *indentWriter) Write(p []byte) (int, error) {
+	iw.beforeWrite()
+	return iw.w.Write(p)
+}
+
+func (iw *indentWriter) WriteString(s string) (int, error) {
+	iw.beforeWrite()
+	return iw.w.WriteString(s)
+}
+
+func (iw *indentWriter) WriteByte(b byte) error {
+	iw.beforeWrite()
+	return iw.w.WriteByte(b)
+}
+
+func (iw *indentWriter) Indent() {
+	iw.currentIndent++
+}
+
+func (iw *indentWriter) Unindent() {
+	iw.currentIndent--
+}
+
+func (iw *indentWriter) NewLine() {
```

Comment:
I wonder whether the calling logic becomes simpler if `NewLine` and `NewLineOrSpace` were removed and the `indentWriter` would automatically apply this logic when a `\n` is encountered in its stream. WDYT?

Reply:
I think it would complicate the code instead of simplifying it. `indentWriter` would need to parse all the strings for `\n`, so the write couldn't just pass the buffer to the underlying writer. On top of that, we have `NewLine` which inserts a new line or nothing and `NewLineOrSpace` which inserts a new line or space. We need to distinguish those two cases, which would mean that we'd need to mark it somehow in the stream, complicating the parsing logic further. So I'd rather keep `NewLine` and `NewLineOrSpace` as they are now.

Reply:
Acknowledged.